### PR TITLE
Write complete manual `QuantumCircuit` documentation

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This writes huge tracts of new `QuantumCircuit` API documentation, linking together alike methods and writing explanatory text for how all the components fit together.  There's likely an awful lot more that could go into this too, but this hopefully should impose a lot more order on the huge `QuantumCircuit` documentation page, and provide a lot more explanation for how the class works holistically.

In particular, the section on the control-flow builder interface could do with a lot more exposition and examples right now.

### Details and comments

I ought to come back and write more about the control-flow builder interface with more examples, but this is already pushing it for 1.1 and I have other things that also need my attention right now.

@Eric-Arellano: at the moment, I've added `QuantumCircuit` to the toctree directly, especially because I don't want it to get generated with the standard templates we use for that.  If that's going to be a problem / not great, we can look at alternatives.

### CODE REVIEW

File: docs/apidoc/qiskit.circuit.QuantumCircuit.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
+.. _qiskit-circuit-quantumcircuit:
+
+==============================
+:class:`.QuantumCircuit` class
```
Comment (Eric-Arellano) [inline docs/apidoc/qiskit.circuit.QuantumCircuit.rst:4]: Some notes if we stick with this rather than the simpler `QuantumCircuit` (no link):

1. We're not consistent with all the other class pages, e.g. https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.AncillaRegister
2. I don't think the link is very useful. On https://qiskit-docs-preview-pr-1380.1799mxdls7qz.us-south.codeengine.appdomain.cloud/api/qiskit/dev/qiskit.circuit.QuantumCircuit, it simply takes you to the class definition which is the first part of the page anyways
3. The styling is messed up for code in headers, tracked by https://github.ibm.com/IBM-Q-Software/iqp-channel-docs/issues/1037. If we stick with this approach, we'll try to fix the issue in the next week

Despite those concerns, I think I agree with your approach here, particularly because it makes the index page more readable with `QuantumCircuit` being escaped as code and having the word `class`. I tried to get a custom title to work in `index.rst` like '\`\`QuantumCircuit\`\` class <qiskit.circuit.QuantumCircuit>', but RST doesn't like the inline formatting with \`.

**Review by Eric-Arellano — COMMENTED** [review]
Nice! Reviewers, check out https://qiskit-docs-preview-pr-1380.1799mxdls7qz.us-south.codeengine.appdomain.cloud/api/qiskit/dev/qiskit.circuit.QuantumCircuit to see how this renders.

--

> @Eric-Arellano: at the moment, I've added QuantumCircuit to the toctree directly, especially because I don't want it to get generated with the standard templates we use for that. If that's going to be a problem / not great, we can look at alternatives.

I think it's good. FYI it results in the index page and ToC looking like this

Screenshot 2024-05-14 at 3 55 10 PM

I'm fine with the index page showing the class since it's so important. And I'm glad the left ToC is still sorted alphabetically. 

Thanks for naming the file `qiskit.circuit.QuantumCircuit` so that we don't need redirects! It would be good to always use that naming scheme moving forward for modules too so that our URLs are more predictable: the import path.

File: docs/apidoc/qiskit.circuit.QuantumCircuit.rst (hunk 1 above)
Comment (jakelishman) [reply docs/apidoc/qiskit.circuit.QuantumCircuit.rst:4]: I mostly did this out of habit because I almost always put in cross-references when possible.  If there's a way that works a little better, I can do that.

From my side, I don't mind the styling not being perfect immediately, if that's the remaining blocker (I'd seen it locally, too). It's a styling bug that we probably need to get fixed anyway, because I'd be surprised if it's not in use in at least some other places - I'm quite likely to have done it somewhere else in the API docs too - but it's still totally legible in the meantime, it just looks slightly odd.

Comment (Eric-Arellano) [reply docs/apidoc/qiskit.circuit.QuantumCircuit.rst:4]: If we could get the `index.rst` to properly render '``QuantumCircuit`` class <qiskit.circuit.QuantumCircuit>', then I would ask us to change to simply `QuantumCircuit`. But I couldn't get that working. So I think what you have is the best choice because it makes the index page more readable.

> From my side, I don't mind the styling not being perfect immediately, if that's the remaining blocker

Indeed, it's my biggest concern. It's always been a problem though, and we'll bump https://github.ibm.com/IBM-Q-Software/iqp-channel-docs/issues/1037 up in priority. It's a short-term bug, so I agree it's not a blocker.

Comment (jakelishman) [reply docs/apidoc/qiskit.circuit.QuantumCircuit.rst:4]: You mean a link where the inline code block and the word "class" are both the link text? If so: nested syntax (inline code inside a link) is (in)famously hard to do in rST. It's just about possible to hack together in some cases, but the syntax is at best arcane.

Sphinx-specific stuff is easier because that produces the nested elements by manipulating the `docutils` tree directly.

Comment (Eric-Arellano) [reply docs/apidoc/qiskit.circuit.QuantumCircuit.rst:4]: > You mean a link where the inline code block and the word "class" are both the link text? If so: nested syntax (inline code inside a link) is (in)famously hard to do in rST. It's just about possible to hack together in some cases, but the syntax is at best arcane.

Yep, exactly. So, I think we should stick with how you have things.

**Comment by jakelishman** [comment]
> Thanks for naming the file `qiskit.circuit.QuantumCircuit` so that we don't need redirects!

Oh, I missed this comment under the picture, but yeah no worries - I'm relatively sure I asked you and Arnau what I should name the file several weeks ago, so I'm just following instructions haha.

File: qiskit/circuit/quantumcircuit.py [hunk 2]
```diff
@@ -106,113 +106,859 @@
...
+
+    The :attr:`name` of a circuit becomes the name of the :class:`~.circuit.Instruction` or
+    :class:`.Gate` resulting from :meth:`to_instruction` and :meth:`to_gate` calls, which can be
+    handy for visualizations.
+
+    .. autoattribute:: name
+
+    You can attach arbitrary :attr:`metadata` to a circuit.  No part of core Qiskit will inspect
+    this or change its behavior based on metadata, but it will be faithfully passed through the
+    transpiler, so you can tag your circuits yourself.
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: Do you want to link to the qpy dump docs here? There are some considerations for qpy circuit serialization around how this gets serialized to json internally.

File: qiskit/circuit/quantumcircuit.py [hunk 3]
```diff
@@ -106,113 +106,859 @@
...
+                              passes.
+    :attr:`parameters`        Ordered set-like view of the compile-time :class:`Parameter`\\ s
+                              tracked by the circuit.
+    :attr:`qregs`             List of :class:`QuantumRegister`\\ s tracked by the circuit.
+
+    :attr:`qubits`            List of :class:`Qubit`\\ s tracked by the circuit.
+    :attr:`unit`              The unit of the :attr:`duration` field.
+    ========================= ======================================================================
+
+    The core attribute is :attr:`data`.  This is a sequence-like object that exposes the
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py:178]: This whole structure is formatted a bit weirdly in: https://qiskit-docs-preview-pr-1380.1799mxdls7qz.us-south.codeengine.appdomain.cloud/api/qiskit/dev/qiskit.circuit.QuantumCircuit the vertical alignment between the sections is weird. I'm wondering if we need to add horizontal lines to break it up, or maybe a lot of subsections. Because of the autoattribute formatting and spacing it looks like the text description below the autoattribute lines are associated with that attribute instead of the text block above it.

(it also does in the rst to me, I kept leaving comments on the wrong spot but figured it would render better)

File: qiskit/circuit/quantumcircuit.py [hunk 4]
```diff
@@ -106,113 +106,859 @@
...
+    You can attach arbitrary :attr:`metadata` to a circuit.  No part of core Qiskit will inspect
+    this or change its behavior based on metadata, but it will be faithfully passed through the
+    transpiler, so you can tag your circuits yourself.
+
+    .. autoattribute:: metadata
+
+    :class:`QuantumCircuit` exposes data attributes tracking its internal quantum and classical bits
+    and registers.  These appear as Python :class:`list`\\ s, but you should treat them as
+    immutable; changing them will *at best* have no effect, and more likely will simply cause
+    :class:`QuantumCircuit`'s internal data to become incoherent.
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: I might reword this as corrupting the structure to make it a bit stronger.

```suggestion
    immutable; changing them will *at best* have no effect, and more likely will simply corrupt
    the internal data of the :class:`QuantumCircuit` object.
```

File: qiskit/circuit/quantumcircuit.py [hunk 5]
```diff
@@ -106,113 +106,859 @@
...
 
-            This argument is intended for convenient circuit initialization when you already have a
-            set of created variables.  The variables used here will be directly passed to
-            :meth:`add_var`, which you can use directly if this is the first time you are creating
-            the variable.
+    Registers are added to the circuit with :meth:`add_register`.  In this method, it is not an
+    error if some of the bits are already present in the circuit.  In this case, the register will
+    be an "alias" over the bits.  This is not generally well-supported by any hardware backends;
+    while it is valid in the OpenQASM 3 abstract data model, it is probably best to stay away from
+    relying on it.  The registers a given bit is in are part of the return of :meth:`find_bit`.
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: I'm not sure we can make an absolute statement about this for all backends. I assume you were referring to IBM backends. Although the non-ibm backends I've worked with don't really model registers at all and just do the tracking in qiskit.

I also don't think we want to couple this with a discussion of the qasm3 data model because it's largely independent from that except if you're exporting the circuit to qasm3.

```suggestion
    be an "alias" over the bits.  This is not generally well-supported by most hardware backends;
    it is probably best to stay away from relying on it.  The registers a given bit is in are part of
     the return of :meth:`find_bit`.
```

File: qiskit/circuit/quantumcircuit.py [hunk 6]
```diff
@@ -106,113 +106,859 @@
...
+
+    Working with bits and registers
+    -------------------------------
+
+    A :class:`.Bit` instance is, on its own, just a unique handle for circuits to use in their own
+    contexts.  If you have got a :class:`.Bit` instance and a cirucit, just can find the contexts
+    that the bit exists in using :meth:`find_bit`, such as its integer index in the circuit and any
+    registers it is contained in.
+
+    .. automethod:: find_bit
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py:375]: I know you haven't had any code examples in the text here. But having an example of using `find_bit` to get the positional index of a bit object either in the ext or part of the automethod (I checked it's not there currently) would be really useful here I think. Doing `qc.find_bit(foo).index` is something I see people needing to do all the time (and I see a lot of people doing `bit._index`)

File: qiskit/circuit/quantumcircuit.py [hunk 7]
```diff
@@ -106,113 +106,859 @@
...
+        :ref:`circuit-adding-data-objects`
+            The methods for adding new :class:`~.expr.Var` variables to a circuit after
+            initialization.
+
+    .. automethod:: get_var
+    .. automethod:: has_var
+    .. automethod:: iter_captured_vars
+    .. automethod:: iter_declared_vars
+    .. automethod:: iter_input_vars
+    .. automethod:: iter_vars
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: Did you want any explainer text between the seealso box and the automethods here?

File: qiskit/circuit/quantumcircuit.py [hunk 8]
```diff
@@ -106,113 +106,859 @@
...
+    :class:`~.circuit.Gate` instances.
+
+    .. seealso::
+        :ref:`circuit-operations-instructions`
+            The :mod:`qiskit.circuit`-level documentation on the different interfaces that Qiskit
+            uses to define circuit-level instructions.
+
+    .. _circuit-append-compose:
+
+    Driver methods to add operations
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: What does driver mean in this context? I would probably have just said "Methods to add operations"

File: qiskit/circuit/quantumcircuit.py [hunk 9]
```diff
@@ -106,113 +106,859 @@
...
+    (:meth:`compose`).
+
+    A single instruction or operation appears as a single entry in the :attr:`data` of the circuit,
+    and as a single box when drawn in the circuit visualizers (see :meth:`draw`).  A single
+    instruction is the "unit" that a hardware backend might be defined in terms of (see
+    :class:`.Target`).  An :class:`~.circuit.Instruction` can come with a
+    :attr:`~.circuit.Instruction.definition`, which is one rule the transpiler (see
+    :mod:`qiskit.transpiler`) will be able to fall back on to decompose it for hardware, if needed.
+    An :class:`.Operation` can only be decomposed if it has some associated high-level synthesis
+    method registered for it (see :mod:`qiskit.transpiler.passes.synthesis.plugin`).
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: I think we should be a bit more clear here, because an `Instruction` and `Gate` object are also `Operation`s. There is also the complexity around opaque instructions and gates which don't include a definition. Which is covered by the previous sentence but not really expanded upon.

```suggestion
    A direct implementer of the :class:`.Operation` interface can only be decomposed if it has some associated high-level synthesis
    method registered for it (see :mod:`qiskit.transpiler.passes.synthesis.plugin`) as it does not include a default
    decomposition via :attr`.Instruction.definition`.
```

File: qiskit/circuit/quantumcircuit.py [hunk 10]
```diff
@@ -106,113 +106,859 @@
...
+    If you are trying to join two circuits that will apply to completely disjoint qubits and clbits,
+    :meth:`tensor` is a convenient wrapper around manually adding bit objects and calling
+    :meth:`compose`.
+
+    .. automethod:: tensor
+
+    As some rules of thumb:
+
+    * If you have a single :class:`.Operation`, :class:`~.circuit.Instruction` or :class:`.Gate`,
+      you should definitely use :meth:`append`.
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: ?
```suggestion
      you should definitely use :meth:`append` or :meth:`_append`.
```

File: qiskit/circuit/quantumcircuit.py [hunk 11]
```diff
@@ -106,113 +106,859 @@
...
+    * :meth:`tensor` is wanted far more rarely than either :meth:`append` or :meth:`compose`.
+      Internally, it is mostly a wrapper around :meth:`add_bits` and :meth:`compose`.
+
+    Some potential pitfalls to beware of:
+
+    * Even if you re-use a custom :class:`~.circuit.Instruction` during circuit construction, the
+      transpiler will generally have to "unroll" each invocation of it to its inner decomposition
+      before beginning work on it.  This should not prevent you from using the
+      :meth:`to_instruction`-plus-:meth:`append` pattern, as the transpiler will also become
+      cleverer about this over time.
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
      :meth:`to_instruction`-plus-:meth:`append` pattern, as the transpiler will improve in this regard
      over time.
```

File: qiskit/circuit/quantumcircuit.py [hunk 12]
```diff
@@ -106,113 +106,859 @@
...
+    .. automethod:: measure_active
+    .. automethod:: measure_all
+
+    There are two "subtractive" methods on :class:`QuantumCircuit` as well.  This is not a use-case
+    that :class:`QuantumCircuit` is designed for; typically you should just look to use
+    :meth:`copy_empty_like` in place of :meth:`clear`, and run :meth:`remove_final_measurements` as
+    its transpiler-pass form :class:`.RemoveFinalMeasurements`.
+
+    .. automethod:: clear
+    .. automethod:: remove_final_measurements
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py:803]: I think showing examples somewhere of the preferred way to do this would be good probably in the automethod docstring.

File: qiskit/circuit/quantumcircuit.py [hunk 13]
```diff
@@ -106,113 +106,859 @@
...
+    directed acyclic graph (DAG).  However, such a definition is a bit hard to grasp, even for
+    experts.  Fortunately, the depth of a circuit can be easily understood by anyone familiar
+    with playing `Tetris <https://en.wikipedia.org/wiki/Tetris>`_.  Lets see how to compute this
+    graphically:
+
+    .. image:: /source_images/depth.gif
+
+    We can verify our graphical result using :meth:`QuantumCircuit.depth`::
+
+       assert qc.depth() == 9
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: I realize this is mostly copied over from before, but I wonder if we want to show an example of 2q gate depth and how to compute that using `depth()`. Either here or in the method docstring.

File: qiskit/circuit/quantumcircuit.py [hunk 14]
```diff
@@ -106,113 +106,859 @@
...
+    which are mostly useful for quickly unwrapping some inner instructions or reversing the
+    :ref:`qubit-labelling conventions <circuit-conventions>` in the drawing.  For more general
+    mutation, including basis-gate rewriting, you should use the transpiler
+    (:mod:`qiskit.transpiler`).
+
+    .. automethod:: decompose
+    .. automethod:: reverse_bits
+
+    ..
+        Deliberately not documented; these methods shouldn't be public.
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: Should we explicitly deprecate them? Since they're in the public api docs we need to support them, but we can make new private methods and switch internal usage over to those and raise a deprecation warnings on the "public" interface.

**Review by mtreinish — COMMENTED** [review]
Overall this looks great, there is a lot much needed detail being added to the docs here which will be a huge help. I have left a few inline comments on some specific minor details. The only big thing I'm concerned about is how this ends up getting formatted on the hosted docs page. The way headings get set for the auto attributes makes it quite confusing, for example:

Screenshot_2024-05-15_13-34-45

All the explainer text looks like it's in the wrong section because of this.

**Comment by Eric-Arellano** [comment]
> All the explainer text looks like it's in the wrong section because of this.

Fwiw, @arnaucasau and I realized this is because the margin-top for h3s is quite large. I started talking to one of the designers if we can reduce it. There's a chance we can reduce it a bit, but they aren't convinced yet we should.

Related context: we went back-and-forth on removing those auto-generated headers and decided to keep them. They're there to populate the right ToC and allow users to copy an anchor link. 

So, the best improvement the docs app could make here is reducing margin-top for h3.

**Comment by jakelishman** [comment]
Eric: I'd say that the styling is actually correct here, the problem is that semantically, a function name etc shouldn't be a document heading. A heading implies that everything after it should be part of that title (which is what the CSS is currently set up to indicate), but that's not semantically what the documentation actually means; the function names are just things to be defined, and that doesn't impact the overall grouping structure of the document.

Sphinx's HTML output is semantically "definition lists" for these, which is (imo) more correct. I know that we currently need headers to make the right sidebar work, but I think that the most proper fix here is to fix the HTML output and then fix whatever knock-on effects that has.

**Comment by Eric-Arellano** [comment]
> Eric: I'd say that the styling is actually correct here, the problem is that semantically, a function name etc shouldn't be a document heading.

You're spot on. I agree, and I appreciate you taking the time to think that through. The problem indeed becomes much more apparent with these custom-written module and class pages, compared to autosummary.

I opened https://github.com/Qiskit/documentation/issues/1395. I still need to get designer sign-off, and it certainly won't be implemented by tomorrow's release, but I hope we can get to it within the next month. We're trying to finish this revamp of API docs before other priorities come up. 

So, from the docs perspective, I wouldn't change anything about this PR re: the weird margins. The docs are written correctly and we'll try to fix this on the app side of things.

File: qiskit/circuit/quantumcircuit.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/circuit/quantumcircuit.py:178]: I'm going to mark this as resolved because of @Eric-Arellano's comment here https://github.com/Qiskit/qiskit/pull/12403#issuecomment-2113575766 and the PR: https://github.com/Qiskit/documentation/issues/1395 which should fix this on the docs hosting side.

File: qiskit/circuit/quantumcircuit.py (hunk 7 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Oh yes, thanks - I must have grouped these into one heading and then forgotten to write the text that goes with them.

File: qiskit/circuit/quantumcircuit.py (hunk 14 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Imo we should (also, this comment was mostly to myself while I was grouping things before I started writing the text), but at this point it's probably best to do that in 1.2.

File: qiskit/circuit/quantumcircuit.py (hunk 13 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: I did copy this (near) verbatim.  I think the text of the section is meant to be read as a whole thing, including the image - the `qc` circuit created up top is the same one as shown in the gif, and it does have a depth of 9.

File: qiskit/circuit/quantumcircuit.py (hunk 12 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py:803]: I mostly did this by explicit see-alsos, but mostly done in f6a2e50f37.

File: qiskit/circuit/quantumcircuit.py (hunk 11 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Done in 956475957f.

File: qiskit/circuit/quantumcircuit.py (hunk 10 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Done in 956475957f.

File: qiskit/circuit/quantumcircuit.py (hunk 9 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Done in 956475957f.

File: qiskit/circuit/quantumcircuit.py (hunk 8 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: I think I meant `append` is the "driver" of `rx`, `cx`, etc.  I'll change it to "methods to add general operations" to contrast with "methods to add standard instructions" below it.  Done in 956475957f.

File: qiskit/circuit/quantumcircuit.py (hunk 7 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Done in bb5456e517.

File: qiskit/circuit/quantumcircuit.py (hunk 6 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py:375]: Added in 36b58ff043.

File: qiskit/circuit/quantumcircuit.py (hunk 5 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Done in 956475957f.

File: qiskit/circuit/quantumcircuit.py (hunk 4 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Done in 956475957f.

File: qiskit/circuit/quantumcircuit.py (hunk 2 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Oh good call - done in 5cc79d724c.

File: qiskit/circuit/quantumcircuit.py (hunk 14 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: In the meantime: I added them back to the documentation in 36b53f4010, and I'll prepare a deprecation PR for 1.2.

Comment (mtreinish) [reply qiskit/circuit/quantumcircuit.py]: I'll open a tracking issue for this so we don't lose the thread

Comment (mtreinish) [reply qiskit/circuit/quantumcircuit.py]: Done here: https://github.com/Qiskit/qiskit/issues/12417

File: qiskit/circuit/quantumcircuit.py (hunk 13 above)
Comment (mtreinish) [reply qiskit/circuit/quantumcircuit.py]: I was thinking more back to some offline conversations we were having with about having a documented way to do 2 qubit depth. But I think this probably belongs as an example in the `depth` method's docstring on how to use the filter arg.

Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: Done in 32a766964d.

**Review by mtreinish — APPROVED** [review]
This LGTM now, thanks for the quick updates.
